File: guides/claude/claude-managed-agents/orchestrator_webhook.py
"""Webhook orchestrator for Claude Managed Agents on Superserve."""
from __future__ import annotations

import logging
import os
import threading

import anthropic
import dotenv
import uvicorn
from fastapi import BackgroundTasks, HTTPException, Request, FastAPI

import orchestrator

dotenv.load_dotenv(override=True)

PORT = int(os.environ.get("PORT", "5051"))
WEBHOOK_SECRET = os.environ.get("ANTHROPIC_WEBHOOK_SECRET")

log = logging.getLogger("orchestrator.webhook")
app = FastAPI()

_client = anthropic.Anthropic(auth_token=orchestrator.ENVIRONMENT_KEY)


def _drain_work() -> None:
    import asyncio
    asyncio.run(_drain_work_async())


async def _drain_work_async() -> None:
    async with anthropic.AsyncAnthropic(auth_token=orchestrator.ENVIRONMENT_KEY) as client:
        async for work in client.beta.environments.work.poller(
            environment_id=orchestrator.ENVIRONMENT_ID,
            environment_key=orchestrator.ENVIRONMENT_KEY,
            block_ms=None,
            reclaim_older_than_ms=int(os.environ.get("POLL_RECLAIM_OLDER_THAN_MS", "2000")),
            drain=True,
            auto_stop=False,
        ):
            try:
                await orchestrator.handle_work(work)
            except Exception as e:
                log.error("failed to handle work=%s: %s", work.id, e, exc_info=True)


@app.post("/")
async def webhook(request: Request, background_tasks: BackgroundTasks):
    if WEBHOOK_SECRET is None:
        raise HTTPException(status_code=500, detail="ANTHROPIC_WEBHOOK_SECRET not set")

    raw = await request.body()
    try:
        event = _client.beta.webhooks.unwrap(
            raw.decode(),
            headers=dict(request.headers),
            key=WEBHOOK_SECRET,
        )
    except anthropic.APIWebhookValidationError as e:
        raise HTTPException(status_code=401, detail=str(e)) from e

    if event.data.type != "session.status_run_started":
        return {"status": "ignored"}

    background_tasks.add_task(_drain_work)
    return {"status": "queued"}


@app.get("/healthz")
def healthz():
    return {"ok": True, "environment_id": orchestrator.ENVIRONMENT_ID}


def main() -> None:
    logging.basicConfig(
        level=os.environ.get("LOG_LEVEL", "INFO"),
        format="%(asctime)s [%(levelname)s] %(name)s: %(message)s",
    )
    if WEBHOOK_SECRET is None:
        raise RuntimeError(
            "ANTHROPIC_WEBHOOK_SECRET is not set. "
            "Use orchestrator.py for polling mode, or set the secret for webhook mode."
        )
    log.info("webhook orchestrator listening on :%d env=%s", PORT, orchestrator.ENVIRONMENT_ID)
    uvicorn.run(app, host="0.0.0.0", port=PORT, log_level="info")


if __name__ == "__main__":
    main()

File: guides/claude/claude-managed-agents/typescript/orchestrator-webhook.mjs
import "dotenv/config"
import Anthropic from "@anthropic-ai/sdk"
import { createServer } from "node:http"
import { handleWork, findOrCreateSandbox } from "./orchestrator.mjs"

const ENVIRONMENT_KEY = process.env.ANTHROPIC_ENVIRONMENT_KEY
const ENVIRONMENT_ID = process.env.ANTHROPIC_ENVIRONMENT_ID
const WEBHOOK_SECRET = process.env.ANTHROPIC_WEBHOOK_SECRET
const PORT = parseInt(process.env.PORT || "5051")

if (!WEBHOOK_SECRET) {
  console.error("ANTHROPIC_WEBHOOK_SECRET is not set")
  process.exit(1)
}

const client = new Anthropic({ authToken: ENVIRONMENT_KEY })
let draining = false

async function drainWork() {
  if (draining) return
  draining = true
  try {
    for await (const work of client.beta.environments.work.poller({
      environmentId: ENVIRONMENT_ID,
      environmentKey: ENVIRONMENT_KEY,
      blockMs: null,
      reclaimOlderThanMs: 2000,
      drain: true,
      autoStop: false,
    })) {
      try {
        await handleWork(work)
      } catch (e) {
        console.error(`failed work=${work.id}: ${e.message}`)
      }
    }
  } finally {
    draining = false
  }
}

function fallbackDrainLoop() {
  setInterval(() => {
    drainWork().catch((e) => console.warn(`fallback drain error: ${e.message}`))
  }, 30_000)
}

const server = createServer(async (req, res) => {
  if (req.method === "GET" && req.url === "/healthz") {
    res.writeHead(200, { "content-type": "application/json" })
    res.end(JSON.stringify({ ok: true, environment_id: ENVIRONMENT_ID }))
    return
  }

  if (req.method !== "POST") {
    res.writeHead(405)
    res.end()
    return
  }

  const chunks = []
  for await (const chunk of req) chunks.push(chunk)
  const body = Buffer.concat(chunks).toString()

  let event
  try {
    event = client.beta.webhooks.unwrap(body, {
      headers: Object.fromEntries(
        Object.entries(req.headers).map(([k, v]) => [k, Array.isArray(v) ? v[0] : v]),
      ),
      key: WEBHOOK_SECRET,
    })
  } catch {
    res.writeHead(401, { "content-type": "application/json" })
    res.end(JSON.stringify({ error: "signature verification failed" }))
    return
  }

  if (event.data.type !== "session.status_run_started") {
    res.writeHead(200, { "content-type": "application/json" })
    res.end(JSON.stringify({ status: "ignored" }))
    return
  }

  drainWork().catch((e) => console.error("drain error:", e.message))

  res.writeHead(200, { "content-type": "application/json" })
  res.end(JSON.stringify({ status: "queued" }))
})

fallbackDrainLoop()

server.listen(PORT, () => {
  console.log(`webhook orchestrator listening on :${PORT} env=${ENVIRONMENT_ID}`)
})
